File: internal/engine/sstable_iterator.go
package engine

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"io"
	"os"
)

type IteratorKV struct {
	InternalKey []byte
	Value       []byte
}

type SSTableIterator struct {
	file   *os.File
	reader *bufio.Reader // FIX 3: Buffered reading to minimize syscalls
	endPos uint32
	curr   uint32 // FIX 1: Local counter to eliminate Seek tax

	lenBuf []byte // Pre-allocated 4-byte buffer for lengths
}

func NewSSTableIterator(path string, endPos uint32) (*SSTableIterator, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}

	return &SSTableIterator{
		file:   f,
		reader: bufio.NewReaderSize(f, 64*1024), // 64KB buffer for smooth streaming
		endPos: endPos,
		curr:   0,
		lenBuf: make([]byte, 4),
	}, nil
}

func (it *SSTableIterator) Next() (IteratorKV, error) {
	// Check against our local counter instead of Seek()
	if it.curr >= it.endPos {
		return IteratorKV{}, io.EOF
	}

	// 1. Read InternalKey Length
	if _, err := io.ReadFull(it.reader, it.lenBuf); err != nil {
		return IteratorKV{}, err
	}
	ikLen := binary.LittleEndian.Uint32(it.lenBuf)
	it.curr += 4

	// 2. FIX 2: Zero-allocation direct read.
	// We allocate the final destination once and read directly into it.
	finalIK := make([]byte, ikLen)
	if _, err := io.ReadFull(it.reader, finalIK); err != nil {
		return IteratorKV{}, fmt.Errorf("failed to read InternalKey: %w", err)
	}
	it.curr += ikLen

	// 3. Read Value Length
	if _, err := io.ReadFull(it.reader, it.lenBuf); err != nil {
		return IteratorKV{}, err
	}
	valLen := binary.LittleEndian.Uint32(it.lenBuf)
	it.curr += 4

	// 4. Read Value
	var finalVal []byte
	if valLen > 0 {
		finalVal = make([]byte, valLen)
		if _, err := io.ReadFull(it.reader, finalVal); err != nil {
			return IteratorKV{}, fmt.Errorf("failed to read Value: %w", err)
		}
		it.curr += valLen
	}

	return IteratorKV{InternalKey: finalIK, Value: finalVal}, nil
}

func (it *SSTableIterator) Close() error {
	return it.file.Close()
}

File: internal/engine2_0/sstable_iterator.cpp
#include "sstable_iterator.hpp"
#include <stdexcept>

static uint32_t Uint32LE(const uint8_t* buf) {
    return static_cast<uint32_t>(buf[0]) |
           (static_cast<uint32_t>(buf[1]) << 8) |
           (static_cast<uint32_t>(buf[2]) << 16) |
           (static_cast<uint32_t>(buf[3]) << 24);
}

SSTableIterator::SSTableIterator(const std::string& path, uint32_t endPos)
    : endPos_(endPos), curr_(0) {
    file_.open(path, std::ios::in | std::ios::binary);
}

SSTableIterator::~SSTableIterator() {
    Close();
}

bool SSTableIterator::Next(IteratorKV& kv) {
    if (curr_ >= endPos_) {
        return false;
    }

    uint8_t lenBuf[4];
    file_.read(reinterpret_cast<char*>(lenBuf), 4);
    if (file_.gcount() != 4) {
        return false;
    }
    uint32_t ikLen = Uint32LE(lenBuf);
    curr_ += 4;

    std::vector<uint8_t> finalIK(ikLen);
    file_.read(reinterpret_cast<char*>(finalIK.data()), ikLen);
    if (file_.gcount() != ikLen) {
        return false;
    }
    curr_ += ikLen;

    file_.read(reinterpret_cast<char*>(lenBuf), 4);
    if (file_.gcount() != 4) {
        return false;
    }
    uint32_t valLen = Uint32LE(lenBuf);
    curr_ += 4;

    std::vector<uint8_t> finalVal(valLen);
    if (valLen > 0) {
        file_.read(reinterpret_cast<char*>(finalVal.data()), valLen);
        if (file_.gcount() != valLen) {
            return false;
        }
        curr_ += valLen;
    }

    kv.InternalKey = std::move(finalIK);
    kv.Value = std::move(finalVal);
    return true;
}

void SSTableIterator::Close() {
    if (file_.is_open()) {
        file_.close();
    }
}
